### PR TITLE
fix(react-calendar-compat): fixed bug with focus disappearance

### LINKED ISSUE #30556: [Bug]: DatePicker closes prematurely after using keyboarding shortcuts
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
@fluentui/react-calendar-compat - "0.0.9"
```

### Are you reporting Accessibility issue?

no

### Reproduction

See bug description

### Bug Description

We are experiencing a bug with the calendar control.

- Open date picker
- Focus the days of the month
- Press "pg up" or "pg dn" on the keyboard
- Note: The month changes, focus is no longer visible
- Press "up" or "down" arrow on the keyboard

## Actual Behavior
The page scrolls and the date picker closes

## Expected Behavior
The selected date moves

See repro here [Compat Components / DatePicker - Default ⋅ Storybook (fluentui.dev)](https://react.fluentui.dev/?path=/docs/compat-components-datepicker--default#week-numbers)

### Logs

_No response_

### Requested priority

Normal

### Products/sites affected

Microsoft Dynamics 365 Business Central

### Are you willing to submit a PR to fix?

no

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by ValentinaKozlova**
## Previous Behavior
Focus disappears after switching months
focus-bug

## New Behavior
focus-fix

- Fixes #30556

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

✅ No changes found
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/Avatar-mount_base.html">624</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/Avatar-mount.html">629</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/Button-mount_base.html">307</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/Button-mount.html">310</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/Field-mount_base.html">1194</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/Field-mount.html">1121</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/FluentProvider-mount_base.html">716</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/FluentProvider-mount.html">835</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/FluentProviderWithTheme-mount_base.html">95</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/FluentProviderWithTheme-mount.html">89</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">34</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">43</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">80</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">74</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/MakeStyles-mount_base.html">880</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/MakeStyles-mount.html">851</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/Persona-mount_base.html">1759</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/Persona-mount.html">1735</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/SpinButton-mount_base.html">1451</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/SpinButton-mount.html">1474</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/SwatchPicker-mount_base.html">1596</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31788/perf-test-react-components/SwatchPicker-mount.html">1632</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: packages/react-components/react-calendar-compat/library/src/components/CalendarDayGrid/CalendarDayGrid.tsx [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
```
Comment (fabricteam) [inline packages/react-components/react-calendar-compat/library/src/components/CalendarDayGrid/CalendarDayGrid.tsx:1]: ## [🕵 fluentuiv9 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31788&commitId=c3b6e06a8e6fd2f1356962d6ede5e95e84d08638&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.badgeMask.normal.chromium.png|0| Added|
|Avatar Converged.basic - Dark Mode.normal.chromium.png|0| Removed|
|Avatar Converged.size+inactive+badge.normal.chromium.png|0| Removed|

</details>


  <details><summary><b>Card Converged</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged.size.normal.chromium.png|0| Removed|

</details>


  <details><summary><b>Card Converged - Interactive</b> 6 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged - Interactive.appearance interactive - Outline.click.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Outline.hover.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Outline.normal.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Subtle.click.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Subtle.hover.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Subtle.normal.chromium.png|0| Added|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

**Review by ling1726 — CHANGES_REQUESTED** [review]
Let's discuss this please, this change will bring _**all of tabster**_ into the public API surface, which would be quite a big development

File: packages/react-components/react-calendar-compat/library/src/components/CalendarDayGrid/CalendarDayGrid.tsx [hunk 2]
```diff
@@ -71,6 +71,25 @@ export const CalendarDayGrid: React.FunctionComponent<CalendarDayGridProps> = pr
   const weeks = useWeeks(props, onSelectDate, getSetRefCallback);
   const animateBackwards = useAnimateBackwards(weeks);
   const [getWeekCornerStyles, calculateRoundedStyles] = useWeekCornerStyles(props);
+  const tableRef = React.useRef<HTMLTableElement>(null);
+  const tabster = useTabster();
+
+  // Hack around tabster setting focus on memoized date missing in DOM after month change
```
Comment (sopranopillow) [inline packages/react-components/react-calendar-compat/library/src/components/CalendarDayGrid/CalendarDayGrid.tsx:77]: Mind adding more info about what's going on and where is focus lost to in the description?

Also would it be possible to not do this with tabster and instead do it with a callback from that button?

Comment (ValentinaKozlova) [reply packages/react-components/react-calendar-compat/library/src/components/CalendarDayGrid/CalendarDayGrid.tsx:77]: Even if it's done as a callback, it's still needs tabster to manage focus

**Comment by ValentinaKozlova** [comment]
> Let's discuss this please, this change will bring _**all of tabster**_ into the public API surface, which would be quite a big development

Probably, it can be fixed inside of a tabster itself. The issue is that tabster is looking for memoized date and sets focus on a wrong element
image

File: packages/react-components/react-calendar-compat/library/src/components/CalendarDayGrid/CalendarGridDayCell.tsx [hunk 3]
```diff
@@ -237,6 +240,7 @@ export const CalendarGridDayCell: React.FunctionComponent<CalendarGridDayCellPro
       tabIndex={isNavigatedDate || isFocusable ? 0 : undefined}
       aria-current={day.isToday ? 'date' : undefined}
       aria-selected={day.isInBounds ? day.isSelected : undefined}
+      {...focusAttributes}
```
Comment (ling1726) [inline packages/react-components/react-calendar-compat/library/src/components/CalendarDayGrid/CalendarGridDayCell.tsx:243]: @mshoho  can you please check `isDefault` would be needed here. The example seems pretty straightforward that this should not be needed

File: packages/react-components/react-calendar-compat/library/src/components/CalendarDayGrid/CalendarDayGrid.tsx (hunk 2 above)
Comment (ValentinaKozlova) [reply packages/react-components/react-calendar-compat/library/src/components/CalendarDayGrid/CalendarDayGrid.tsx]: Fixed in a different way. Hope this fix is fine

**Review by sopranopillow — APPROVED** [review]
nice fix! I think you should still wait for Marat to answer, but I'll approve :)

**Comment by ValentinaKozlova** [comment]
Waiting for this fix to be merged. Than this PR won't be needed: https://github.com/microsoft/tabster/pull/393

**Comment by sopranopillow** [comment]
> Waiting for this fix to be merged. Than this PR won't be needed: [microsoft/tabster#393](https://github.com/microsoft/tabster/pull/393)

should this PR be close then? I can block it to avoid accidentally merging if you would like

**Comment by ValentinaKozlova** [comment]
> > Waiting for this fix to be merged. Than this PR won't be needed: [microsoft/tabster#393](https://github.com/microsoft/tabster/pull/393)
> 
> should this PR be close then? I can block it to avoid accidentally merging if you would like

Yeah, I'll close it

**Pull request closed without merging**